### ISSUE TITLE
custom authorizer: callback accepts "string" data but fails with it

### ISSUE THREAD

**Issue by nros** [pusher/pusher-js#427]
This is a bug in the Typescript type definition or in its implementation.

**What is the current behavior?**
Custom authorizer function receive a callback that require two parameters. A boolean value indicating an error and the authentication data. For TypeScript the type of "data" is set to `AuthInfo` (see file [`core/auth/options.ts`](https://github.com/pusher/pusher-js/blob/3e49245daea296f97b8ab439a7a9efa3a4367fa8/src/core/auth/options.ts#L15)).

`AuthInfo` can either be a `string` an object of type `AuthData`. However, the callback can only work with `AuthData` and will fail with type `string` (see file [src/core/channels/channel.ts](https://github.com/pusher/pusher-js/blob/3e49245daea296f97b8ab439a7a9efa3a4367fa8/src/core/channels/channel.ts#L108)).

**If the current behavior is a bug, please provide the steps to reproduce and
if possible a minimal demo of the problem via https://jsfiddle.net or similar.**

* use custom authorization function
* provide a `string` value to the callback

**What is the expected behavior?**
The call back should check for string type and then convert the string value to the `AuthData` instance itself.

**Which versions of Pusher, and which browsers / OS are affected by this issue?
Did this work in previous versions of Pusher? If so, which?**

at least 5.1 is affected, I did not test with earlier versions.

**Comment by leesio** [comment]
@nros, I see that this is not idea. 
Unfortunately the `AuthorizerCallback` is used slightly inconsistently internally in the library - if the first argument (indicating an error) is `true` the second argument passed is a string describing the error detail e.g. https://github.com/pusher/pusher-js/blob/0d718adf06566766e0f544d74565d2ff3dcc0ff8/src/runtimes/isomorphic/auth/xhr_auth.ts#L37-L41

 I know there are several ways we could improve this, my favourite at the moment (subject to change 😄 ) is to encode the error into the `AuthInfo` type.

```
export interface AuthData {
  auth: string;
  channel_data?: string;
  error:? boolean;
  errorDetail:? string;
}
```

However, this would be a breaking change, and I wanted to focus on releasing declarations that describe the current state.

Interested in your thoughts on how you'd like to use this API

**Comment by nros** [comment]
Thinking about it, I see three possible way of handling it.

* silently convert and maybe write a warning to the console
* throw an exception if the data is of string type and but error flag is `false`. May problem was, that I did not see the error at first, because it was hidden in the websocket message. The WS message stated, that auth data is needed, but did not indicate, that it was provided in wrong form.
* use breaking changes. I would prefer to use an `Error` instance instead of a boolean. Than we have both, the indication, that an error occurred, and the error message. Something like:
    ```TypeScript
    type AuthorizerCallback = (data: Error | AuthData) => void;
    ```
    It would be possible to rewrite the method for all JavaScript only users, to support both - new and old - version. TypeScript users would see a compile time error, though - until changed to new types. (below is totally untested - just my thoughts!)
    ```JavaScript
    this.authorize(this.pusher.connection.socket_id, (error, data) => {
      if (typeof error === "boolean" && error) {
        // Why not bind to 'pusher:subscription_error' a level up, and log there?
        // Binding to this event would cause the warning about no callbacks being
        // bound (see constructor) to be suppressed, that's not what we want.
        Logger.error(data);
        this.emit('pusher:subscription_error', data);

      } else if (error instanceof Error) {
        Logger.error(error.message);
        this.emit('pusher:subscription_error', error.message);

      } else {
        if (data === undefined && typeof error === "object" && error !== null) {
          data = error;
        }

        if (!data || data.auth === undefined || data.auth === null) {
            var excp = new Error("auth callback is called with wrong parameters by custom authorizer.");
            console.error(excp);
            throw excp;
        }

        data = data as AuthData;
        this.pusher.send_event('pusher:subscribe', {
          auth: data.auth,
          channel_data: data.channel_data,
          channel: this.name
        });
      }
    });
    ```

**Comment by nros** [comment]
@lessio any progress? Did you have some time to consider my approach?

**Comment by leesio** [comment]
Hey @nros, I'm picking this up again today. Sorry for the delay.

Looking at the api again this morning. I'm not sure why we even send a `bool` indicating when there's been an error. I think the typical behaviour in this scenario is to include the `error` string as the first argument in the callback and have the data as the second.

It would change the 
```
// old
export type AuthorizerCallback = (error: boolean, authInfo: AuthInfo) => void;
```

```
// new
export type AuthorizerCallback = (error: Error | string, authInfo: AuthData) => void
```

That's still a breaking change but I think it's probably the most idiomatic thing to do. Any thoughts on that approach?

Btw: sorry for missing your message above - you misspelled my github username so I didn't get a notification 🙈

**Comment by nros** [comment]
@leesio using a `string` value as error message is better than using a `boolean`. Overall your proposal follows suit other systems I have seen, like nodeJS callbacks. Nevertheless, I would pass the error instance directly, if there is any. It provides a message but also the stack trace. So, receiver can use both.

Anyway, you proposal is an improvement which I would be happy about.

Is there anything I can do?

**Comment by leesio** [comment]
@nros thanks for the offer of help, you can certainly try out the branch with the PR I intend to merge and check it works for you.

**Comment by leesio** [comment]
@nros I just published release 7.0.0 which includes this change. Please let me know if this solves your problem!
